### PR TITLE
Multithread safe settings

### PR DESCRIPTION
**Pull request by andyneff**
Mitigate the new issues with ThreadPoolExecutor

### CODE REVIEW

File: terra/executor/thread.py [hunk 1]
```diff
@@ -1,9 +1,39 @@
...
 
 
 class ThreadPoolExecutor(concurrent.futures.ThreadPoolExecutor,
                          terra.executor.base.BaseExecutor):
-  pass
+  '''
+  Terra version of :class:`concurrent.futures.ThreadPoolExecutor`
+
+  Unlike other executors, :class:`ThreadPoolExecutor` has no process isolation.
+  This results in a scenario where multiple threads could be terra settings,
```
Comment (scott-vsi) [inline terra/executor/thread.py:15]: This results in a scenario where multiple threads using terra settings could actually influence

File: terra/executor/thread.py [hunk 2]
```diff
@@ -1,9 +1,39 @@
...
+  structure.
+
+  This behavior is limited to threads started by ThreadPoolExecutor only. All
+  other threads will have normal thread behavior with the runner threads, and
+  use a single version of the settings. The only side effect is if a task
+  starts its own thread, it will be treated as one of the runner threads, not a
+  task thread. The currently known downside to this is log messages will be
+  reported as coming from the runner rather than task zone. However, any
+  attempts to edit settings from this rouge thread can have unintended
+  consequences.
```
Comment (scott-vsi) [inline terra/executor/thread.py:31]: This seems ominous

File: terra/executor/thread.py [hunk 3]
```diff
@@ -1,9 +1,39 @@
...
+  This results in a scenario where multiple threads could be terra settings,
+  actually influence each other, which is not typical behavior, given that
+  all other executors have process isolation and do not allow this.
+
+  :class:`ThreadPoolExecutor` will downcast :obj:`terra.core.settings` to a
+  thread-safe :class:`terra.core.settings.LazySettingsThreaded` where each
+  Executor thread has it's own thread local storage version of the settings
+  structure.
+
+  This behavior is limited to threads started by ThreadPoolExecutor only. All
```
Comment (scott-vsi) [inline terra/executor/thread.py:24]: :class:`ThreadPoolExecutor`

File: terra/executor/thread.py [hunk 4]
```diff
@@ -1,9 +1,39 @@
...
+  other threads will have normal thread behavior with the runner threads, and
+  use a single version of the settings. The only side effect is if a task
+  starts its own thread, it will be treated as one of the runner threads, not a
+  task thread. The currently known downside to this is log messages will be
+  reported as coming from the runner rather than task zone. However, any
+  attempts to edit settings from this rouge thread can have unintended
```
Comment (NoahRJohnson) [inline terra/executor/thread.py:30]: rouge -> rogue

File: terra/executor/thread.py (hunk 1 above)
Comment (andyneff) [reply terra/executor/thread.py:15]: That was supposed to read `This results in a scenario where multiple threads could use terra settings to influence each other`

**Pull request merged**